gitpod-io/gitpod
# [server] Allow to login with an instllation admin user

## Description

This PR introduces a single, per-installation `admin-user` with a predefined userId, which can login to a web session using an OTS that needs to be created using a private endpoint beforehand. 

- add a static `admin-user` that is `blocked` by default
- adds a non-exposed endpoint to create an OTS token for login purposes for that `admin-user` (`/admin-user/login-token/create`)
- adds an exposed endpoint to login into a web session using said token (`/login/ots/admin-user/:key`)
- integrate with preview deployment: if `with-dedicated-emulation` is ticked, output a fresh OTS with login link in werft

### Login flow

:information_source: Be aware this is meant as a _first version_ to get us going with _something_. Depending on product requirements (ability to re-login), and technical possibilities/constraints from Dedicated Cells (Scout Team) this might to change again in the actual implementation, especially around token generation/handling.

Current flow impl:
1. installer generates a `admin-login-key` secret that gets mounted into `server`
1. on `/admin-user/login-token/create`, `server` creates a new OTS (expiry: 24h), where the hash is sugared with that secret. Also, the `admin-user` stays `blocked` until the first call here
2. on `/login/ots/admin-user/:key`, `server` validates the OTS value using that secret, and puts the client into a web session with `admin-user`

## Related Issue(s)

Fixes #15657

## How to test

1. Start [workspace on this branch](https://gitpod-io/#https://github.com/gitpod-io/gitpod/pull/15805) and `kubectl port-forward deployment/server 9000:9000 &`
2. Obtain OTS key: `curl -X POST localhost:9000/admin-user/login-token/create`
3. Use the result to construct your login URL: `https://gpl-15657-setup-user.preview.gitpod-dev.com/api/login/ots/admin-user/:yourOtsKey` and visit it
4. You should be logged in as `admin-user` :checkered_flag: 


## Release Notes

```release-note
NONE
```

## Documentation


## Build Options:

      Experimental feature to run the build with GitHub Actions (and not in Werft).
      If enabled this will build `install/preview`
- [ ] /werft with-dedicated-emulation
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
